Upgrade from Ogre 1.10 to Ogre 1.12.1

Martin-Idel opened the pull request:
This PR pulls upgrades to the latest OGRE 1.x version, pulling in a number of breaking API changes.

Let's see whether this helps with https://github.com/ros2/rviz/issues/368.

This may or may not help with issues on OSX (maybe if we iterate a bit, this may help with #385 - but I don't have OSX and some fixes may be necessary to build Ogre on Mojave: https://github.com/OGRECave/ogre/issues/894).

Commits are hopefully structured to make reviewing this easier.

Martin-Idel committed:
Update Ogre to 1.12.1

- old boost patch seems unnecessary

Martin-Idel committed:
Fix OgreVector3 now being a typedef

Martin-Idel committed:
Replace deprecated function getAttachedObjectsIterator

Martin-Idel committed:
Replace destroySceneNode with new named version

Martin-Idel committed:
Fix font definition file

- Could not be loaded anymore

Martin-Idel committed:
Minor cleanup in include of ros_image_texture

Martin-Idel committed:
Load Codec for png images (now a plugin)

Martin-Idel commented:
Sadly, this doesn't help for #368 :

* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3059)](http://ci.ros2.org/job/ci_linux-aarch64/3059/)

I can start a full CI if you'd like to perform the upgrade.

Martin-Idel committed:
Replace deprecated method getChildIterator

VictorLamoine commented:
I know I'm late to the party but why don't we use Ogre 2?

Martin-Idel commented:
@VictorLamoine 

Ogre 2 uses a different API - in part its vastly different. This effort was deemed too much in the beginning of the migration and it would probably (?) change the RViz plugins API and require remigration of plugins. The benefits are mostly only in terms of performance on modern architectures.

Since Ogre 1 is (for now) still maintained and actively developed (in part because it is used by gazebo and RViz), we can stay on Ogre 1.

However, out of curiosity I'm investigating what it would actually mean to upgrade to Ogre 2 and I hope to be able to show this soonish. 

Note also that `rviz_rendering` was originally designed to encapsulate all Ogre usage to be able to swap out the implementation at some point. It turned out that Ogre is (as of now) used in too many aspects throughout the code base - especially in the plugin API so that this effort was not pursued further.

mjcarroll commented:
@iche033 could probably provide some more input on what changes may be needed, as he has been going through this with ign-rendering, for use with the next-generation gazebo.

I know that the shaders were an issue, and had to be modified for OGRE2.

iche033 commented:
From the top of my head, here are the changes I had to do when porting to use ogre 2.x for ign-rendering 

* upgrade old ogre 1.x materials to the ogre 2.x's HLMS. 
    - the old materials still exist - at least on the API level but I found that some functionalities are no longer available so I upgraded them all to HLMS.
* upgrading glsl shaders to version >150
* update engine initialization and resource setup
    - this is more straightforward. 
* convert cameras to use compositors
* fix various API breakages

Other changes:
* I had to change the way selection buffer (for mouse picking) is implemented since material schemes are not supported anymore.
* We're using ogre RTSS in gazebo and that's no longer available so another reason that I updated the materials to use the PBS component of HLMS.
 * I spend quite a bit of time integrating ogre2.x with QML as it's not as trivial as integration with QWidgets. Yet there are still issues on macOS (and very likely on Windows too).

I'm not very familiar with rviz code base but a quick look shows that things like movable texts, and point clouds may also need to be updated since it's using the old material system.

Martin-Idel commented:
Thanks for the information. That was about what I expected and then some :-D.

Anyway, until then I rebased this PR and updated it to the latest release (no changes).

Martin-Idel committed:
Disable a few more components which cause CMake errors

Martin-Idel commented:
@wjwwood , @jacobperron 
As far as I can see, the latest changes to 1.12.1 (or maybe even to 1.12?) actually make the tests pass on aarch in repeated mode:

* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3862)](https://ci.ros2.org/job/ci_linux-aarch64/3862/)

So this PR seems to finally fix #368 - I know that tests are currently disabled in the nightlies so they could be reenabled once this is merged.

wjwwood commented on rviz_common/test/interaction/mock_selection_renderer.hpp:
```diff
@@ -94,7 +94,7 @@ class MockSelectionRenderer : public rviz_common::interaction::SelectionRenderer
     dst_box = Ogre::PixelBox(width, height, 1, Ogre::PF_A8R8G8B8, data);
 
-    auto dstptr = static_cast<uint32_t *>(dst_box.data);
+    auto dstptr = reinterpret_cast<uint32_t *>(dst_box.data);
```
What's the purpose of this change?

Martin-Idel replied:
Sadly the `dst_box.data` type was changed from array of `uint32_t` capturing the four bytes per pixel in one int to array of `uchar` for each pixel. The layout of the data is the same which is why the `reinterpret_cast` works. I chose to first do it that way to show that I'm not really altering anything - the test is the same as before.

It's probably better to rework the setup to use `uint8_t` and I can do it if you want.

wjwwood replied:
I would like to see either a comment explaining that here (that we're interpreting it as four items in a 32-bit integer), or to see it refactored. I don't mind which though.

Martin-Idel replied:
I left a comment. Note that I figured out that the change in data structure means that some `reinterpret_cast<uint8_t *>` in the `selection_manager` and `view_picker` are now superfluous. I changed them to `static_cast` to a) help explain what we are expecting at these points and b) have something in the code that'd complain if the data structure were changed again in some future update.

wjwwood commented on rviz_rendering/src/rviz_rendering/ogre_render_window_impl.cpp:
```diff
@@ -123,7 +122,8 @@ RenderWindowImpl::render()
   // experience it seems better to keep things separate and keep the render
   // function as simple as possible.
-  Ogre::WindowEventUtilities::messagePump();
+  // TODO(Martin-Idel): This should be removed
```
Should this be resolved before merging? Why is the messagePump no longer required?

Martin-Idel replied:
Of course, I just forgot about it. It was moved to `OgreBites`, which I disabled because it needs SDL to build completely and throws a CMake warning if you don't. Also, this would be the one and only place where we need `OgreBites` so I tried to just disable it and never found anything wrong with it.

To be honest, I don't know what the message pump is actually supposed to do. If you look at the implementation (https://github.com/OGRECave/ogre/blob/f4709042cf7722757d86fa830fedea0985017d53/Components/Bites/src/OgreWindowEventUtilities.cpp) you'll see:
- it does nothing on Mac
- It does something on Linux, but it iterates over render windows which have to be registered. Since the corresponding function was never called, I don't think it does anything on Linux (and I couldn't see any change in functionality).
- It might do something on Windows (three lines that I'm inclined to copy if they are actually needed), but then again I don't see any difference.

wjwwood replied:
The Windows part seems relevant, as it seems to be processing GUI "MSG's" or something. I have no idea if we're using those, or if Qt is pumping them for us or not. Similarly the Linux code seems to be processing messages from the OS about change of focus and occlusion.

This line seems to indicate that on Windows it avoids the "not responding" behavior (happens when a program freezes):

https://github.com/OGRECave/ogre/blob/1fdf49ce312ee8f1c4426f826326601aaef1b521/Components/Bites/src/OgreApplicationContextBase.cpp#L550

Which makes sense that an application needs to do something for Windows to know whether it is or isn't responding. But again, I'd expect Qt does this for us?

---

I spoke with @iche033 about it and I think this function might be doing some useful things like disabling rendering of the render window when the window is occluded:

https://github.com/OGRECave/ogre/blob/f4709042cf7722757d86fa830fedea0985017d53/Components/Bites/src/OgreWindowEventUtilities.cpp#L335-L353

But I think it's only meant to be used by Ogre applications (demos and samples), and instead we could/should use the Qt events in a similar fashion.

I'm not sure we need to do that now, but perhaps remove this and make an issue pointing out that our behavior may have changed since we no longer call this function, and that we should consider emulating some of its behavior with Qt events instead, as follow up work.

---

Also, there's this issue: https://github.com/ros2/rviz/issues/328

But I don't think its directly related. I'm not sure if it is speaking about losing focus within a Qt widget or like losing focus of a window or something.

Martin-Idel replied:
#328 is definitely a Qt-Problem (the focus of the display widget is lost), so that's not related.

I did a bit more digging and also found https://forums.ogre3d.org/viewtopic.php?t=66358 where it says 
>If you are using Qt, it also pumps messages indiscriminately of window. You shouldn't need to call WindowEventUtilities::messagePump() unless your Qt window and Ogre window are on separate threads

which is in line with what you are saying and is not contradicted by other OGRE devs in the same thread [but is contradicted by the official wiki at https://wiki.ogre3d.org/Integrating+Ogre+into+QT5 which I guess was the original inspiration].

For now, I'll leave a comment and will create an issue later.

wjwwood replied:
Sounds good to me! Thanks for looking into it.

Martin-Idel committed:
Fix Overlay build

- upstream Ogre now uses Find modules that CMake provides
- However, this resulted in freetype not being found

Martin-Idel committed:
Delete messagePump and leave comment

Martin-Idel committed:
Explain reinterpret_cast for PixelBox

Martin-Idel committed:
Change reinterpret casts to static casts

The underlying data structure changed so we don't
need the reinterpret casts anymore.
Leaving the static_casts to make explicit what we
expect at that point and force action if the model
changes again on some update

Martin-Idel commented:
I rebased the branch again and added commits starting with one commit fixing the `Ogre::Overlay` build on Windows for me.

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7799)](http://ci.ros2.org/job/ci_linux/7799/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3876)](http://ci.ros2.org/job/ci_linux-aarch64/3876/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6365)](http://ci.ros2.org/job/ci_osx/6365/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7632)](http://ci.ros2.org/job/ci_windows/7632/)

I'll consider re enabling the tests that are disabled on nightly if these pass.

Martin-Idel commented:
So it's "just" a few thousand Windows warnings and otherwise it works. My first try to get rid of them seems to not work on my machine, so I'll have to think of a true fix probably...

Martin-Idel commented:
It's just one warning in many iterations but as far as I can see, there is no true fix. However, the code is completely safe as the problem the warning C4251 describes cannot occur: we compile both Ogre and RViz with the same compiler. 

The only way to get rid of the warnings would be to disable them. There are three ways, which would you prefer?

- disable the warning via CMake (I'd prefer this, but couldn't get it to work, just adding `add_compile_options(/wd4251)` for MSVC-compiler didn't do the trick)
- add pragmas to virtually all OGRE headers via a patch
- add pragmas around many Ogre includes in RViz files

---
Here is why I think the warning cannot be fixed:

The new templated `Ogre::Vector` class instantiates a few members (such as `Vector3::ZERO`, etc.) inside the library while the template definition is also provided and exposed in the header file. This means that when I use e.g. `Ogre::Vector3` in RViz, I will generate my own vector instantiation while I use the one in Ogre's library when I call `Ogre::Vector3::ZERO`. That's exactly the situation which generates the C4251 warning and what it's for.

I thought one could maybe explicitly instantiate all vectors needed for RViz within Ogre, but that is not possible because the Ogre code relies on implicit (lazy) template instantiation - for instance, it defines a three argument constructor for all vectors, but then provides a specialization for 2-dim vectors which doesn't have a three argument constructor so that when I explicitly instantiate a 2-dim vector, I get an error about the missing three argument constructor...

Martin-Idel committed:
Remove old pragmas for MSVC

Martin-Idel committed:
Cleanup GCC pragmas

Martin-Idel committed:
Add pragma patch to Ogre

Martin-Idel committed:
Remove more MSVC pragmas

Martin-Idel committed:
Remove probably obsolete clang pragmas

Martin-Idel committed:
Fix include order after pragma removal

Martin-Idel committed:
Rename project to reflect that we are building 1.12.1

Martin-Idel commented:
@wjwwood I somehow couldn't disable the warning C4251 via CMake, something seems to override the settings. If you have a better idea, I'm happy to try that.

For now, I added pragmas to Ogre via a patch (making this PR quite large, but most of it is a diff-file) using `sed` scripts. As described above, I don't see a way to actually fix the warnings. The patch is only applied on Windows.

Since I was on the task, I also cleaned up a bit: All old Ogre pragmas for other warnings are gone since they are no longer necessary and I also changed the project name of the ogre `ExternalProject`.

I took the liberty to redo CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7863)](http://ci.ros2.org/job/ci_linux/7863/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3925)](http://ci.ros2.org/job/ci_linux-aarch64/3925/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6417)](http://ci.ros2.org/job/ci_osx/6417/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7724)](http://ci.ros2.org/job/ci_windows/7724/)

wjwwood commented:
I'm ok with the patch approach. Thanks for looking into it, I know disabling those warnings is a PITA when it's on Windows and an external project.
